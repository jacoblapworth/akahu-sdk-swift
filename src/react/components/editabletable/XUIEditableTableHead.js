import React from 'react';
import PropTypes from 'prop-types';

import cn from 'classnames';

import { tableName } from './private/constants';

const baseName = `${tableName}head`;

const XUIEditableTableHead = ({ children, className, ...spreadProps }) => {
  return (
    <div className={cn(baseName, className)} {...spreadProps}>
      {children}
    </div>
  );
};

XUIEditableTableHead.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
};

export default XUIEditableTableHead;
