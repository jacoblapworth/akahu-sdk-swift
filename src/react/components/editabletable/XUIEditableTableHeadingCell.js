import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import { tableName } from './private/constants';

const baseName = `${tableName}headingcell`;

const XUIEditableTableHeadingCell = ({ children, className, qaHook, scope, ...spreadProps }) => {
  // TODO: sort out how weâ€™re going to handle text alignment and add some logic here
  const cellClassName = cn(baseName, className, `${baseName}-leftaligned`);

  return (
    <th className={cellClassName} data-automationid={qaHook} scope={scope} {...spreadProps}>
      {children}
    </th>
  );
};

XUIEditableTableHeadingCell.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  qaHook: PropTypes.string,
  /**
   * The `scope` attribute added to the <th> element
   * to tell screenreaders exactly what cells the header is a header for
   * <br />
   * Default value is `col`, could be `colGroup` / `row` / `rowGroup`
   */
  scope: PropTypes.oneOf(['col', 'colGroup', 'row', 'rowGroup']),
};

export default XUIEditableTableHeadingCell;

XUIEditableTableHeadingCell.defaultProps = {
  scope: 'col',
};
