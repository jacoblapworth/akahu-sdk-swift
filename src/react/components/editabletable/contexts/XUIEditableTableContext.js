import React from 'react';

export default React.createContext({
  rowOptions: {
    isRemovable: false,
  },
});
