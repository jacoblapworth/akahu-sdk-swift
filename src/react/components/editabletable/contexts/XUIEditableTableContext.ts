import * as React from 'react';

interface XUIEditableTableContext {
  columnWidths?: string[];
  dragAndDrop: {
    dragHandleDescribedBy?: string;
    rowPlaceholder?: React.ReactElement<HTMLElement> | null;
  };
  rowOptions: {
    dragButtonAriaLabel?: string;
    isDraggable?: boolean;
    isRemovable?: boolean;
    removeButtonAriaLabel?: string;
  };
  scrollContainerRef?: React.MutableRefObject<HTMLElement>;
  tableRef?: React.MutableRefObject<HTMLTableElement>;
  tableWrapperRef?: React.MutableRefObject<HTMLElement>;
}

export default React.createContext<XUIEditableTableContext>({
  dragAndDrop: {},
  rowOptions: {},
});
