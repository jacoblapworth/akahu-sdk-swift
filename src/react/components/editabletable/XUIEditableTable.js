import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import { tableName } from './private/constants';
import XUIEditableTableContext from './contexts/XUIEditableTableContext';
import XUIIcon from '../icon/XUIIcon';
import invalid from '@xero/xui-icon/icons/invalid';
import generateIds from '../controlwrapper/helpers';

const XUIEditableTable = ({
  caption,
  children,
  className,
  columnWidths = [],
  isInvalid,
  maxWidth,
  minWidth,
  qaHook,
  rowOptions,
  style,
  validationMessage,
  ...spreadProps
}) => {
  const tableRef = React.useRef();
  const wrapperStyle =
    // If we omit this check, wrapperStyle is always a non-empty object, and passes extraneous (but harmless) props.
    // This is for tidiness purposes only.
    maxWidth || minWidth
      ? {
          maxWidth,
          minWidth,
          ...style,
        }
      : style;
  const wrapperIds = generateIds(spreadProps.id);

  return (
    <div className={cn(`${tableName}--wrapper`, className)}>
      <div
        className={cn(tableName, isInvalid && `${tableName}-is-invalid`)}
        ref={tableRef}
        style={wrapperStyle}
      >
        <table
          {...spreadProps}
          aria-describedby={isInvalid && validationMessage ? wrapperIds.message : undefined}
          aria-invalid={isInvalid}
          className={`${tableName}--table`}
          data-automationid={qaHook}
        >
          {caption && <caption className={`${tableName}--caption`}>{caption}</caption>}
          {!!columnWidths.length && (
            <colgroup>
              {columnWidths.map(item => (
                <col style={{ width: item }} />
              ))}
              {rowOptions.isRemovable && <col style={{ width: '40px' }} />}
              {/* 40px is $xui-control-size-standard */}
            </colgroup>
          )}
          <XUIEditableTableContext.Provider value={{ rowOptions: { ...rowOptions }, tableRef }}>
            {children}
          </XUIEditableTableContext.Provider>
        </table>
      </div>
      {isInvalid && validationMessage && (
        <div className={`${tableName}--validation`} id={wrapperIds.message}>
          <XUIIcon icon={invalid} />
          {validationMessage}
        </div>
      )}
    </div>
  );
};

XUIEditableTable.propTypes = {
  /**
   * A non-visible description of the table for accessibility purposes. Particularly useful
   * for scrollable tables, to help screenreaders understand the scrollable element.
   */
  caption: PropTypes.string,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
  className: PropTypes.string,
  qaHook: PropTypes.string,
  rowOptions: PropTypes.shape({
    isRemovable: PropTypes.bool,
    removeButtonAriaLabel(props, propName, componentName) {
      if (!props.isRemovable) {
        return null;
      }

      if (!props[propName]) {
        return new Error(
          `The prop \`rowOptions.${propName}\` is required by \`${componentName}\` when using \`rowOptions.isRemovable\`, but its value is \`undefined\`.`,
        );
      }

      if (typeof props[propName] !== 'string') {
        return new Error(
          `Invalid prop \`${propName}\` of type \`${typeof props[
            propName
          ]}\` supplied to \`${componentName}\`, expected \`string\`.`,
        );
      }

      return null;
    },
  }),
  columnWidths: PropTypes.arrayOf(PropTypes.string),
  /**
   * Used to style the table as invalid.
   */
  isInvalid: PropTypes.bool,
  maxWidth: PropTypes.string,
  minWidth: PropTypes.string,
  style: PropTypes.object,
  /**
   * Validation message to show under the table if `isInvalid` is true.
   */
  validationMessage: PropTypes.string,
};

export default XUIEditableTable;
