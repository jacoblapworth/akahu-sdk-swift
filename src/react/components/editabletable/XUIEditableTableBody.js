import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';

import XUIEditableTableContext from './contexts/XUIEditableTableContext';
import XUIEditableTableClassContext from './contexts/XUIEditableTableClassContext';

const XUIEditableTableBody = ({ children, className, qaHook, ...spreadProps }) => {
  const {
    dragAndDrop: { rowPlaceholder },
  } = React.useContext(XUIEditableTableContext);
  const tableClassName = React.useContext(XUIEditableTableClassContext);
  const baseName = `${tableClassName}body`;

  return (
    <tbody className={cn(baseName, className)} data-automationid={qaHook} {...spreadProps}>
      {children}
      {rowPlaceholder}
    </tbody>
  );
};

XUIEditableTableBody.propTypes = {
  children: PropTypes.node,
  className: PropTypes.string,
  qaHook: PropTypes.string,
};

export default XUIEditableTableBody;
