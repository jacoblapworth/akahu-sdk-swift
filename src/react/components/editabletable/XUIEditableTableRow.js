import React from 'react';
import PropTypes from 'prop-types';
import cn from 'classnames';
import trashIcon from '@xero/xui-icon/icons/trash';

import { tableName } from './private/constants';
import XUIEditableTableHeadingCell from './XUIEditableTableHeadingCell';
import XUIEditableTableCellIconButton from './XUIEditableTableCellIconButton';
import XUIEditableTableContext from './contexts/XUIEditableTableContext';
import XUIEditableTableHeadContext from './contexts/XUIEditableTableHeadContext';

const baseName = `${tableName}row`;

// TODO: Replace this with $xui-control-size-standard when SASS tokens can be shared with React
const xuiControlSizeStandard = '40px';

const XUIEditableTableRow = ({ children, className, onRemove, qaHook, ...spreadProps }) => {
  const isHeaderRow = React.useContext(XUIEditableTableHeadContext);
  const {
    rowOptions: { isRemovable, removeButtonAriaLabel },
  } = React.useContext(XUIEditableTableContext);

  return (
    <tr className={cn(baseName, className)} data-automationid={qaHook} {...spreadProps}>
      {children}
      {isRemovable && isHeaderRow && (
        <XUIEditableTableHeadingCell style={{ width: xuiControlSizeStandard }} />
      )}
      {isRemovable && !isHeaderRow && (
        <XUIEditableTableCellIconButton
          ariaLabel={removeButtonAriaLabel}
          iconReference={trashIcon}
          onClick={onRemove}
          qaHook={`${qaHook}--button-remove`}
        />
      )}
    </tr>
  );
};

XUIEditableTableRow.propTypes = {
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.arrayOf(PropTypes.element)]),
  className: PropTypes.string,
  onRemove: PropTypes.func,
  qaHook: PropTypes.string,
};

export default XUIEditableTableRow;
