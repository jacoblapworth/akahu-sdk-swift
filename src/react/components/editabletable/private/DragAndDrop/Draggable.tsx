import * as PropTypes from 'prop-types';
import * as React from 'react';
import {
  Draggable as RBDDraggable,
  DraggableProvided,
  DraggableRubric,
  DraggableStateSnapshot,
} from 'react-beautiful-dnd';
import { Portal } from 'react-portal';
import { v4 as uuid } from 'uuid';

import XUIEditableTableContext from '../../contexts/XUIEditableTableContext';
import { xuiControlSizeStandard } from '../constants';
import DragDropDraggingContext from './contexts/DragDropDraggingContext';
import DroppableContext from './contexts/DroppableContext';
import addOffsetToTransform from './helpers/addOffsetToTransform';

interface BaseProps {
  children(
    provided?: Omit<DraggableProvided, 'draggableProps'> & {
      draggableProps: Partial<
        Omit<DraggableProvided['draggableProps'], 'style'> & {
          style: React.CSSProperties;
        }
      >;
    },
    snapshot?: DraggableStateSnapshot,
    rubric?: DraggableRubric,
  ): React.ReactElement<HTMLElement>;
  draggableId?: string;
  index?: number;
  useDraggable?: boolean;
}

type Props = BaseProps &
  Omit<React.ComponentProps<typeof RBDDraggable>, 'children' | 'draggableId' | 'index'>;

/**
 * Draggable is a wrapper around (and has the same API as) react-beautiful-dnd's
 * [Draggable](https://github.com/atlassian/react-beautiful-dnd/blob/master/docs/api/draggable.md).
 *
 * General changes this wrapper makes:
 * - `draggableId` is automatically generated if it is not provided
 *
 * Table-oriented changes this wrapper makes:
 * - Modify the preview to look like the draggable will go back to its origin when it is dragged
 *   outside of the droppable.
 * - The draggable's transform is modified to account for the table's border-spacing
 * - While being dragged the draggable is wrapped in a `table` element that keeps track of column
 *   widths
 */
const Draggable: React.FunctionComponent<Props> = ({
  children,
  draggableId,
  index,
  useDraggable,
  ...spreadProps
}) => {
  const [generatedDraggableId] = React.useState(uuid());

  const { draggedRowHeight, draggedRowIndex } = React.useContext(DragDropDraggingContext);
  const { isDraggingOver } = React.useContext(DroppableContext);
  const { columnWidths, rowOptions, tableRef } = React.useContext(XUIEditableTableContext);

  if (!useDraggable || typeof index !== 'number') {
    return children();
  }

  return (
    <RBDDraggable draggableId={draggableId || generatedDraggableId} index={index} {...spreadProps}>
      {(provided, snapshot, rubric) => {
        /**
         * Table-oriented change 1/3
         *
         * Modify the preview to look like the draggable will go back to its origin when it is
         * dragged outside of the droppable.
         */
        const makeRoomForDraggable =
          !isDraggingOver && typeof draggedRowIndex === 'number' && draggedRowIndex < index;

        const providedTransform =
          provided.draggableProps.style?.transform ||
          (makeRoomForDraggable && `translate(0px, ${draggedRowHeight}px)`) ||
          'translate(0px, 0px)';

        /**
         * Table-oriented change 2/3
         *
         * Modify the draggable's transform to account for the table's border-spacing.
         */
        const transform = addOffsetToTransform(providedTransform, snapshot.isDragging);

        if (!snapshot.isDragging) {
          return children(
            {
              ...provided,
              draggableProps: {
                ...provided.draggableProps,
                style: {
                  ...provided.draggableProps.style,
                  pointerEvents: isDraggingOver ? 'none' : undefined,
                  transform,
                },
              },
            },
            snapshot,
            rubric,
          );
        }

        /**
         * Table-oriented change 3/3
         *
         * Wrap the draggable in a `table` element that keeps track of column widths.
         */
        const columns = Array.from(tableRef?.current?.querySelector('tr')?.children || []) as Array<
          HTMLElement
        >;
        const calculatedColumnWidths = columns
          .filter(
            (_, columnIndex) =>
              !(
                (rowOptions.isSortable && columnIndex === 0) ||
                (rowOptions.isRemovable && columnIndex === columns.length - 1)
              ),
          )
          .map(cell => `${cell.offsetWidth}px`);

        const columnWidthsToUse = calculatedColumnWidths?.length
          ? calculatedColumnWidths
          : columnWidths;

        return (
          <Portal>
            <table
              {...provided.draggableProps}
              style={{
                ...provided.draggableProps.style,
                transform,
                // Copy the styling of XUIEditableTable
                background: 'white',
                borderSpacing: '1px',
                borderCollapse: 'separate',
                tableLayout: 'fixed',
              }}
            >
              {!!columnWidthsToUse?.length && (
                <colgroup>
                  {rowOptions.isSortable && <col style={{ width: xuiControlSizeStandard }} />}
                  {columnWidthsToUse.map((width, i) => (
                    <col key={String(i)} style={{ width }} />
                  ))}
                  {rowOptions.isRemovable && <col style={{ width: xuiControlSizeStandard }} />}
                </colgroup>
              )}
              <tbody>
                {children(
                  {
                    ...provided,
                    draggableProps: {
                      style: {
                        transition: provided.draggableProps.style?.transition,
                      },
                    },
                  },
                  snapshot,
                  rubric,
                )}
              </tbody>
            </table>
          </Portal>
        );
      }}
    </RBDDraggable>
  );
};

Draggable.propTypes = {
  children: PropTypes.func.isRequired,
  draggableId: PropTypes.string,
  index: PropTypes.number,
  useDraggable: PropTypes.bool,
};

export default Draggable;
