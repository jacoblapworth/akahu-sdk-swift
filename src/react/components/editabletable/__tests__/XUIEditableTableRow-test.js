import React from 'react';
import Enzyme, { mount, shallow } from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import toJson from 'enzyme-to-json';
import { nanoid } from 'nanoid';
import { axe, toHaveNoViolations } from 'jest-axe';

import XUIEditableTable from '../XUIEditableTable';
import XUIEditableTableRow from '../XUIEditableTableRow';
import XUIEditableTableBody from '../XUIEditableTableBody';
import XUIEditableTableHead from '../XUIEditableTableHead';
import XUIEditableTableHeadingCell from '../XUIEditableTableHeadingCell';
import Draggable from '../private/DragAndDrop/Draggable';
import NOOP from '../../helpers/noop';
import XUIEditableTableContext from '../contexts/XUIEditableTableContext';

Enzyme.configure({ adapter: new Adapter() });
expect.extend(toHaveNoViolations);

jest.mock('nanoid');
nanoid.mockImplementation(() => 'testRowId');

jest.mock('../private/DragAndDrop/Draggable');
Draggable.mockImplementation(({ children }) => children());

describe('<XUIEditableTableRow />', () => {
  it('renders correctly', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <XUIEditableTableRow>
            <td>XUIEditableTableRow</td>
          </XUIEditableTableRow>
        </tbody>
      </table>,
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('composes the className correctly', () => {
    const wrapper = mount(
      <table>
        <tbody>
          <XUIEditableTableRow className="test-classname" />
        </tbody>
      </table>,
    );
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('should pass accessibility testing', async () => {
    const wrapper = mount(
      <table>
        <tbody>
          <XUIEditableTableRow className="test-classname" />
        </tbody>
      </table>,
    );
    const results = await axe(wrapper.html());
    expect(results).toHaveNoViolations();
  });

  describe('removable row', () => {
    it('renders a remove button for rows in the body', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: true, removeButtonAriaLabel: 'Remove row' }}>
          <XUIEditableTableBody>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row--button-remove"]').length).toBe(1);
    });

    it('renders a blank table cell for rows in the head', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: true, removeButtonAriaLabel: 'Remove row' }}>
          <XUIEditableTableHead>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableHead>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableHeadingCell).length).toBe(1);
    });

    it('calls the onRemove callback when the remove button is clicked', () => {
      // Arrange
      const onRemoveMock = jest.fn();
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: true, removeButtonAriaLabel: 'Remove row' }}>
          <XUIEditableTableBody>
            <XUIEditableTableRow onRemove={onRemoveMock} qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      // Act
      wrapper.find('[data-automationid="test-row--button-remove"]').simulate('click');

      // Assert
      expect(onRemoveMock).toBeCalled();
    });

    it('should disable the remove icon when `disableRowControls` prop is true', () => {
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: true, removeButtonAriaLabel: 'Remove row' }}>
          <XUIEditableTableBody>
            <XUIEditableTableRow qaHook="test-row" disableRowControls />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      expect(
        wrapper
          .find('[data-automationid="test-row--button-remove"]')
          .hasClass('xui-button-is-disabled'),
      ).toBeTruthy();
    });
  });

  describe('non-removable row', () => {
    it('does not render a remove button for rows in the body', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: false }}>
          <XUIEditableTableBody>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row--button-remove"]').length).toBe(0);
    });

    it('does not render a blank table cell for rows in the head', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isRemovable: false }}>
          <XUIEditableTableHead>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableHead>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableHeadingCell).length).toBe(0);
    });
  });

  describe('draggable row', () => {
    it('renders a drag button for rows in the body', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable
          dndDragCancelledMessage={NOOP}
          dndDragOutsideMessage={NOOP}
          dndDragStartMessage={NOOP}
          dndDragUpdateMessage={NOOP}
          dndDropFailedMessage={NOOP}
          dndDropMessage={NOOP}
          dndInstructions=""
          onReorderRow={NOOP}
          rowOptions={{ isDraggable: true, dragButtonAriaLabel: 'Drag row' }}
        >
          <XUIEditableTableBody>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row--button-drag"]').length).toBe(1);
    });

    it('renders a blank table cell for rows in the head', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable
          dndDragCancelledMessage={NOOP}
          dndDragOutsideMessage={NOOP}
          dndDragStartMessage={NOOP}
          dndDragUpdateMessage={NOOP}
          dndDropFailedMessage={NOOP}
          dndDropMessage={NOOP}
          dndInstructions=""
          onReorderRow={NOOP}
          rowOptions={{ isDraggable: true, dragButtonAriaLabel: 'Drag row' }}
        >
          <XUIEditableTableHead>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableHead>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableHeadingCell).length).toBe(1);
    });

    it('spreads provided.draggableProps onto the table row', () => {
      // Arrange
      const draggableProps = { test: 'test' };
      Draggable.mockImplementation(({ children }) => children({ draggableProps }));

      const wrapper = mount(
        <table>
          <tbody>
            <XUIEditableTableRow qaHook="test-row">
              <td>XUIEditableTableRow</td>
            </XUIEditableTableRow>
          </tbody>
        </table>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row"]').props().test).toStrictEqual(
        draggableProps.test,
      );
    });

    it('spreads provided.dragHandleProps onto the drag cell', () => {
      // Arrange
      const dragHandleProps = { test: 'test' };
      Draggable.mockImplementation(({ children }) => children({ dragHandleProps }));

      const wrapper = mount(
        <XUIEditableTableContext.Provider
          value={{
            dragAndDrop: { dragHandleDescribedBy: '' },
            rowOptions: { dragButtonAriaLabel: '', isDraggable: true },
          }}
        >
          <table>
            <tbody>
              <XUIEditableTableRow qaHook="test-row">
                <td>XUIEditableTableRow</td>
              </XUIEditableTableRow>
            </tbody>
          </table>
        </XUIEditableTableContext.Provider>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row--cell-drag"]').props().test).toStrictEqual(
        dragHandleProps.test,
      );
    });

    it('merges props.style with provided.draggableProps.style', () => {
      // Arrange
      const providedDraggablePropsStyle = { background: 'purple' };
      const propsStyle = { color: 'purple' };
      const expectedStyle = { ...providedDraggablePropsStyle, ...propsStyle };

      Draggable.mockImplementation(({ children }) =>
        children({
          draggableProps: { style: providedDraggablePropsStyle },
        }),
      );

      const wrapper = mount(
        <table>
          <tbody>
            <XUIEditableTableRow style={propsStyle} qaHook="test-row">
              <td>XUIEditableTableRow</td>
            </XUIEditableTableRow>
          </tbody>
        </table>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row"]').prop('style')).toStrictEqual(
        expectedStyle,
      );
    });

    it('should disable the drag icon when `disableRowControls` prop is true', () => {
      const wrapper = mount(
        <XUIEditableTable
          dndDragCancelledMessage={NOOP}
          dndDragOutsideMessage={NOOP}
          dndDragStartMessage={NOOP}
          dndDragUpdateMessage={NOOP}
          dndDropFailedMessage={NOOP}
          dndDropMessage={NOOP}
          dndInstructions=""
          onReorderRow={NOOP}
          rowOptions={{ isDraggable: true, dragButtonAriaLabel: 'Drag row' }}
        >
          <XUIEditableTableBody>
            <XUIEditableTableRow disableRowControls qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      expect(
        wrapper
          .find('[data-automationid="test-row--button-drag"]')
          .hasClass('xui-button-is-disabled'),
      ).toBeTruthy();
    });

    it('disables dragging when `disableRowControls` prop is true', () => {
      const wrapper = mount(
        <XUIEditableTable
          dndDragCancelledMessage={NOOP}
          dndDragOutsideMessage={NOOP}
          dndDragStartMessage={NOOP}
          dndDragUpdateMessage={NOOP}
          dndDropFailedMessage={NOOP}
          dndDropMessage={NOOP}
          dndInstructions=""
          onReorderRow={NOOP}
          rowOptions={{ isDraggable: true, dragButtonAriaLabel: 'Drag row' }}
        >
          <XUIEditableTableBody>
            <XUIEditableTableRow disableRowControls qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      expect(wrapper.find(Draggable).prop('isDragDisabled')).toBeTruthy();
    });
  });

  describe('non-draggable row', () => {
    it('does not render a drag button for rows in the body', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isDraggable: false }}>
          <XUIEditableTableBody>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableBody>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find('[data-automationid="test-row--button-drag"]').length).toBe(0);
    });

    it('does not render a blank table cell for rows in the head', () => {
      // Arrange
      const wrapper = mount(
        <XUIEditableTable rowOptions={{ isDraggable: false }}>
          <XUIEditableTableHead>
            <XUIEditableTableRow qaHook="test-row" />
          </XUIEditableTableHead>
        </XUIEditableTable>,
      );

      // Assert
      expect(wrapper.find(XUIEditableTableHeadingCell).length).toBe(0);
    });
  });
});
