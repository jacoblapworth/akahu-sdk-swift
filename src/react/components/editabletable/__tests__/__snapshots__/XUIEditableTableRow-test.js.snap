// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<XUIEditableTableRow /> composes the className correctly 1`] = `
<tr
  className="xui-editabletablerow test-classname"
/>
`;

exports[`<XUIEditableTableRow /> removable row renders a remove button for rows in the body 1`] = `
<XUIEditableTable
  rowOptions={
    Object {
      "isRemovable": true,
      "removeButtonAriaLabel": "Remove row",
    }
  }
>
  <div
    className="xui-editabletable"
  >
    <table
      className="xui-editabletable--table"
    >
      <XUIEditableTableBody>
        <tbody
          className="xui-editabletablebody"
        >
          <XUIEditableTableRow
            qaHook="test-row"
          >
            <tr
              className="xui-editabletablerow"
              data-automationid="test-row"
            >
              <XUIEditableTableCellIconButton
                ariaLabel="Remove row"
                iconReference={
                  Object {
                    "height": 14,
                    "path": "M0 1h3l1-1h3l1 1h3v2H0V1zm1 3h9v9.455c0 .272-.346.545-.692.545H1.692C1.346 14 1 13.727 1 13.455V4zm2 2v6h1V6H3zm2 0v6h1V6H5zm2 0v6h1V6H7z",
                    "width": 11,
                  }
                }
                qaHook="test-row--button-remove"
              >
                <XUIEditableTableCell
                  className="xui-editabletablecelliconbutton"
                >
                  <td
                    className="xui-editabletablecell xui-editabletablecelliconbutton"
                  >
                    <XUIIconButton
                      ariaLabel="Remove row"
                      icon={
                        Object {
                          "height": 14,
                          "path": "M0 1h3l1-1h3l1 1h3v2H0V1zm1 3h9v9.455c0 .272-.346.545-.692.545H1.692C1.346 14 1 13.727 1 13.455V4zm2 2v6h1V6H3zm2 0v6h1V6H5zm2 0v6h1V6H7z",
                          "width": 11,
                        }
                      }
                      isDisabled={false}
                      isExternalLink={false}
                      isLink={false}
                      qaHook="test-row--button-remove"
                      tabIndex={0}
                      type="button"
                    >
                      <XUIButton
                        aria-label="Remove row"
                        className="xui-button-icon-medium xui-button-icon"
                        fullWidth="never"
                        isDisabled={false}
                        isExternalLink={false}
                        isGrouped={false}
                        isLink={false}
                        isLoading={false}
                        qaHook="test-row--button-remove"
                        retainLayout={true}
                        tabIndex={0}
                        type="button"
                        variant="unstyled"
                      >
                        <button
                          aria-label="Remove row"
                          className="xui-button xui-button-icon-medium xui-button-icon"
                          data-automationid="test-row--button-remove"
                          disabled={false}
                          onClick={[Function]}
                          tabIndex={0}
                          type="button"
                        >
                          <XUIIcon
                            icon={
                              Object {
                                "height": 14,
                                "path": "M0 1h3l1-1h3l1 1h3v2H0V1zm1 3h9v9.455c0 .272-.346.545-.692.545H1.692C1.346 14 1 13.727 1 13.455V4zm2 2v6h1V6H3zm2 0v6h1V6H5zm2 0v6h1V6H7z",
                                "width": 11,
                              }
                            }
                            role="presentation"
                            size="medium"
                          >
                            <svg
                              className="xui-icon"
                              focusable="false"
                              height={14}
                              viewBox="0 0 11 14"
                              width={11}
                            >
                              <path
                                d="M0 1h3l1-1h3l1 1h3v2H0V1zm1 3h9v9.455c0 .272-.346.545-.692.545H1.692C1.346 14 1 13.727 1 13.455V4zm2 2v6h1V6H3zm2 0v6h1V6H5zm2 0v6h1V6H7z"
                                role="presentation"
                              />
                            </svg>
                          </XUIIcon>
                          <XUITouchTarget>
                            <div
                              className="xui-touchtarget"
                            />
                          </XUITouchTarget>
                        </button>
                      </XUIButton>
                    </XUIIconButton>
                  </td>
                </XUIEditableTableCell>
              </XUIEditableTableCellIconButton>
            </tr>
          </XUIEditableTableRow>
        </tbody>
      </XUIEditableTableBody>
    </table>
  </div>
</XUIEditableTable>
`;

exports[`<XUIEditableTableRow /> renders correctly 1`] = `
<tr
  className="xui-editabletablerow"
>
  <div>
    XUIEditableTableRow
  </div>
</tr>
`;
